Add SOP on how to update the observability operator CR manually (#88)
RHCloudServices/kafka-monitoring-stuff: sops/update_observability_operator_cr.asciidoc

// begin header
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]
:numbered:
:toc: macro
:toc-title: pass:[<b>Table of Contents</b>]
// end header
= Manually updating the Observability Operator CR on an OSD cluster

toc::[]

== Description

> The purpose of this SOP is to update Observability operator CR on an OSD cluster after the initial installation of the operator was done.  

Reviewer: Can we add something here that explains why this has to be done manually?
Currently it reads like this has to always be done after the initial installation of the operator (which I believe isn't the case. Just for currently existing clusters or whenever the CR changes again in future and needs to be updated again in existing clusters)

> The purpose of this SOP is to update Observability operator CR on an OSD cluster after the initial installation of the operator was done.  

Reviewer 2: ```suggestion
The purpose of this SOP is to update an existing observability operator installation (OLM Bundles, observability CR etc) on a data plane OSD cluster since the Managed Services API does not reconcile changes for existing installations.  
```

> This is done to update the configuration of the operator e.g to update secrets of the alert manager.  

Reviewer 2: ```suggestion
```

This makes sure that the operator configurations are up to date for its end to end functionality. 
In the future, this step will likely be done automatic (see https://issues.redhat.com/browse/MGDSTRM-882), but that depends on how the observability operator will be installed. 

== Prerequisites

> 1. The target OSD cluster exists. Note the `cluster-id`.

Reviewer: What is the target OSD cluster?
Any/all clusters where the 'data plane' is running?

(I realise there's only 1 data plane cluster at present, but this SOP could be around for a while until https://issues.redhat.com/browse/MGDSTRM-882 is done, so I'd like to make it as clear as possible)

> 2. The Observability operator has been installed on the target cluster via the syncset approach. The id of the syncset could be `ext-managed-application-services-observability` (Prior to Jan, 2021) or `ext-managedservice-cluster-mgr` (as of Jan, 2021) depending on when the cluster was terraformed.

Reviewer: How to verify if the operator was installed this way (via syncset)?

Author: This will be via the `get` command given in the execution step.

> The commands in this SOP assumes that the id is `ext-managed-application-services-observability`.

Reviewer: If the id is not what's expected, should it be updated to match?
Is it a problem if it doesn't match?

Author: The id could be one or the other. The current id is `ext-managedservice-cluster-mgr` but this only applies to new clusters, but the stage cluster is still using the old id `ext-managed-application-services-observability`.
The syncset will always have one or the other.

3. The https://github.com/openshift-online/ocm-cli[`ocm`] CLI.

> 4. Cluster credentials are held that allow execution of `ocm` commands. 

Reviewer: There's a lot of different permissions that can be applied to OCM accounts.
What permissions or roles are needed specifically here?

> 4. Cluster credentials are held that allow execution of `ocm` commands. 

Reviewer 2: The individual running this SOP would also need permissions to see the OSD clusters created in the organisation used by the Managed Services API.

Author: Added.

5. The https://stedolan.github.io/jq/[`jq`] CLI

== Execute/Resolution

> 

Reviewer 2: Would be good to set an env var for the cluster ID and SyncSet ID and just use them in the below commands. E.g.

`syncset-id=ext-managed-application-services-observability` and then just reference in the below commands as `$syncset-id`

Author: Thanks. I'll update this.

1. Login to the `ocm` CLI against the target environment (`stage` or `prod`)
2. Execute the following to retrieve the content of the existing operator CR:

```

> ocm get /api/clusters_mgmt/v1/clusters/<cluster-id>/external_configuration/syncsets/ext-managed-application-services-observability > cr.json

Reviewer: My understanding is a little lacking here.
I thought syncsets were created in the same cluster as where the hive operator is running.
Does this API do that for you?

Reviewer 2: that is correct @david-martin . What is referred to as the trusted datapath are the syncset endpoints exposed through OCM cluster service which give the ability to interact with SyncSets on Hive.

```

NOTE: change the `cluster-id` with the id value of the target cluster.

4. Upon inspection of the content, you will see an `Observability` resource, with contents like the one below
```
{
    "kind": "Observability",
    "spec": {
    "grafana": {
        "managed": false
    },
    "observatorium": {
        "tenant": "tenant-value",
        "gateway": "https://example.io",
        "authType": "dex",
        "dexConfig": {
        "url": "http://example.com",
        "credentialSecretName": "example-secret",
        "credentialSecretNamespace": "namespace"
     }
    },
    "canaryPodSelector": {
        "matchLabels": {
        "managed-kafka-canary": "true"
        }
    },
    "kafkaNamespaceSelector": {
        "matchLabels": {
        "mas-managed": "true"
        }
    }
    },
    "metadata": {
    "name": "observability-stack",
    "namespace": "managed-application-services-observability",
    "creationTimestamp": null
    },
    "apiVersion": "observability.redhat.com/v1"
}
```

> You can change the Observability kind resource to update the value of the existing contents or add new fields. 

Reviewer: Can this step be made easier?
The link is to the CRD, which is quite verbose and more likely to result in human error when parsing it.
Also, what values should be updated, new fields added and existing values left as is specifically?

Author: What's need to be updated is the content of the `cr.json` file. The CRD is only linked to see the schema. 
The values to be updated depends on what we want to update - the previous two cases it has been secrets but I did not want to call any specific fields as all fields can be updated.

Reviewer: Hmm, OK.
So the fields (and values) to update would be communicated as needed whenever this SOP has to be followed?

Author: > So the fields (and values) to update would be communicated as needed whenever this SOP has to be followed?

Yes. WDYT?  Do you want this to be called out in the SOP?

Reviewer: Good idea. Maybe something like `The fields and values to be modified should be communicated from the person or team requesting this SOP to be followed` ?

That would hint at the SOP potentially being needed many times

> The content of the CR has to match the https://github.com/bf2fc6cc711aee1a0c2a/observability-operator/blob/c6e8d9be4de1bec6d597ef7abfe8131d05bd5b36/config/crd/bases/observability.redhat.com_observabilities.yaml[Observability operator CRD]

Reviewer 3: the link should be inside parentheses to be rendered as an URL:

)https://github.com/bf2fc6cc711aee1a0c2a/observability-operator/blob/c6e8d9be4de1bec6d597ef7abfe8131d05bd5b36/config/crd/bases/observability.redhat.com_observabilities.yaml)[Observability operator CRD]

Once the modifications of the CR are done, save the file and execute the following command to update the syncset.

```

> jq 'del(.id)' cr.json | ocm patch /api/clusters_mgmt/v1/clusters/<cluster-id>/external_configuration/syncsets/ext-managed-application-services-observability

Reviewer 3: what's the `jq 'del(.id)'` for?

Author: This remove the id from the `cr.json` content. This is because the content of this cr are saved from the get command above.

> jq 'del(.id)' cr.json | ocm patch /api/clusters_mgmt/v1/clusters/<cluster-id>/external_configuration/syncsets/ext-managed-application-services-observability

Reviewer 2: ```suggestion
jq 'del(.id)' cr.json | ocm patch /api/clusters_mgmt/v1/clusters/<cluster-id>/external_configuration/syncsets/ext-managed-application-services-observability --body=cr.json
```

```

== Validate

1. Run the following command and ensure that the syncset has been updated.

```
ocm get /api/clusters_mgmt/v1/clusters/<cluster-id>/external_configuration/syncsets/ext-managed-application-services-observability
```

> 2. Login in to the cluster and verify that the `managed-application-services-observability` 

Reviewer: Login to which cluster?
The OSD cluster?

Author: The OSD cluster.

I've added this clarification.

In the future, this step will likely be done automatic (see https://issues.redhat.com/browse/MGDSTRM-882), but that depends on how the observability operator will be installed. 

== Prerequisites

1. The target OSD cluster exists.
2. The Observability operator has been installed on the target cluster via the syncset approach. The id of the syncset could be `ext-managed-application-services-observability` (Prior to Jan, 2021) or `ext-managedservice-cluster-mgr` (as of Jan, 2021) depending on when the cluster was terraformed.
The commands in this SOP assumes that the id is `ext-managed-application-services-observability`.
3. The https://github.com/openshift-online/ocm-cli[`ocm`] CLI.
4. Cluster credentials are held that allow execution of `ocm` commands.
The individual running this SOP would also need permissions to see the OSD clusters created in the organisation used by the Managed Services API.
5. The https://stedolan.github.io/jq/[`jq`] CLI

== Execute/Resolution

1. Login to the `ocm` CLI against the target environment (`stage` or `prod`)
2. Execute the following to retrieve the content of the existing operator CR:

```
syncsetId=ext-managed-application-services-observability

> clusterId=target-cluster-id

Reviewer: Let's call out the target cluster id as a prereq,
and also include an example cmd on how to retrieve or determine this.

Author: Good point. I'll add it.

        "matchLabels": {
        "mas-managed": "true"
        }
    }
    },
    "metadata": {
    "name": "observability-stack",
    "namespace": "managed-application-services-observability",
    "creationTimestamp": null
    },
    "apiVersion": "observability.redhat.com/v1"
}
```

You can change the Observability kind resource to update the value of the existing contents or add new fields. 
The content of the CR has to match the https://github.com/bf2fc6cc711aee1a0c2a/observability-operator/blob/c6e8d9be4de1bec6d597ef7abfe8131d05bd5b36/config/crd/bases/observability.redhat.com_observabilities.yaml[Observability operator CRD]
Once the modifications of the CR are done, save the file and execute the following command to update the syncset.

```
jq 'del(.id)' cr.json | ocm patch /api/clusters_mgmt/v1/clusters/$clusterId/external_configuration/syncsets/$syncsetId --body=cr.json
```

== Validate

1. Run the following command and ensure that the syncset has been updated.

```
ocm get /api/clusters_mgmt/v1/clusters/$clusterId>/external_configuration/syncsets/$syncsetId
```

2. Login in to the OSD cluster and verify that the `managed-application-services-observability` 

> is there and the observability operator is running with the correct configuration. 

Reviewer: How to verify the operator is running with the correct configuration?
Is it different depending on the fields changed?
If so, are there a couple of examples that could be put here?

Author: I'll add a command to check the the observability CR has been updated to match the modified values.
